src/jdk.internal.le/linux/native/lible/CLibrary.cpp

```diff
@@ -0,0 +1,206 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "jni.h"
+#include "jni_util.h"
+#include "jvm.h"
+#include "jdk_internal_org_jline_terminal_impl_jna_linux_CLibraryImpl.h"
+
+#include <errno.h>
+#include <stdlib.h>
+#include <termios.h>
+#include <unistd.h>
+#include <sys/ioctl.h>
+
+static jclass lastErrorExceptionClass;
+static jmethodID lastErrorExceptionConstructor;
+
+static jclass termios_j;
+static jfieldID c_iflag;
+static jfieldID c_oflag;
+static jfieldID c_cflag;
+static jfieldID c_lflag;
+static jfieldID c_line;
+static jfieldID c_cc;
+static jfieldID c_ispeed;
+static jfieldID c_ospeed;
+
+static jclass winsize_j;
+static jfieldID ws_row;
+static jfieldID ws_col;
+static jfieldID ws_xpixel;
+static jfieldID ws_ypixel;
+
+JNIEXPORT void JNICALL Java_jdk_internal_org_jline_terminal_impl_jna_linux_CLibraryImpl_initIDs
+  (JNIEnv *env, jclass) {
+    jclass cls;
+    cls = env->FindClass("jdk/internal/org/jline/terminal/impl/jna/LastErrorException");
+    CHECK_NULL(cls);
+    lastErrorExceptionClass = (jclass) env->NewGlobalRef(cls);
+    lastErrorExceptionConstructor = env->GetMethodID(lastErrorExceptionClass, "<init>", "(J)V");
+    CHECK_NULL(lastErrorExceptionConstructor);
+
+    cls = env->FindClass("jdk/internal/org/jline/terminal/impl/jna/linux/CLibrary$termios");
+    CHECK_NULL(cls);
+    termios_j = (jclass) env->NewGlobalRef(cls);
+    c_iflag = env->GetFieldID(termios_j, "c_iflag", "I");
+    CHECK_NULL(c_iflag);
+    c_oflag = env->GetFieldID(termios_j, "c_oflag", "I");
+    CHECK_NULL(c_oflag);
+    c_cflag = env->GetFieldID(termios_j, "c_cflag", "I");
+    CHECK_NULL(c_cflag);
+    c_lflag = env->GetFieldID(termios_j, "c_lflag", "I");
+    CHECK_NULL(c_lflag);
+    c_line = env->GetFieldID(termios_j, "c_line", "B");
+    CHECK_NULL(c_line);
+    c_cc = env->GetFieldID(termios_j, "c_cc", "[B");
+    CHECK_NULL(c_cc);
+    c_ispeed = env->GetFieldID(termios_j, "c_ispeed", "I");
+    CHECK_NULL(c_ispeed);
+    c_ospeed = env->GetFieldID(termios_j, "c_ospeed", "I");
+    CHECK_NULL(c_ospeed);
+
+    cls = env->FindClass("jdk/internal/org/jline/terminal/impl/jna/linux/CLibrary$winsize");
+    CHECK_NULL(cls);
+    winsize_j = (jclass) env->NewGlobalRef(cls);
+    ws_row = env->GetFieldID(winsize_j, "ws_row", "S");
+    CHECK_NULL(ws_row);
+    ws_col = env->GetFieldID(winsize_j, "ws_col", "S");
+    CHECK_NULL(ws_col);
+    ws_xpixel= env->GetFieldID(winsize_j, "ws_xpixel", "S");
+    CHECK_NULL(ws_xpixel);
+    ws_ypixel= env->GetFieldID(winsize_j, "ws_ypixel", "S");
+    CHECK_NULL(ws_ypixel);
+}
+
+JNIEXPORT void JNICALL Java_jdk_internal_org_jline_terminal_impl_jna_linux_CLibraryImpl_tcgetattr
+  (JNIEnv *env, jobject, jint fd, jobject result) {
+    termios data;
+
+    if (tcgetattr(fd, &data) != 0) {
+        jobject exc = env->NewObject(lastErrorExceptionClass,
```

Comment:
Could lines 103-106 be converted to a utility function which would then be invoked at current lines 103, 141, 163, and 197?